File: Leetcode75/1029-vertical-order-traversal-of-a-binary-tree/vertical-order-traversal-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalTraversal(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        column_table = defaultdict(list)
        queue = deque([(root, 0, 0)])  # node, column index, row index

        while queue:
            node, column, row = queue.popleft()
            if node:
                column_table[column].append((row, node.val))  # append (row, value)
                queue.append((node.left, column - 1, row + 1))  # go left
                queue.append((node.right, column + 1, row + 1))  # go right

        # Sort the column indices and prepare the result
        sorted_columns = sorted(column_table.keys())
        result = []
        for col in sorted_columns:
            # Sort by row first, then by value
            column_nodes = sorted(column_table[col], key=lambda x: (x[0], x[1]))
            result.append([val for row, val in column_nodes])  # extract values

        return result

File: Leetcode75/1029-vertical-order-traversal-of-a-binary-tree/vertical-order-traversal-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // Helper function to perform DFS
void dfs(TreeNode* node, int row, int col, map<int, priority_queue<tuple<int, int>, vector<tuple<int, int>>, greater<>>>& nodes) {
    if (!node) return;

    // Add the node to the map with its row and value
    nodes[col].emplace(row, node->val);

    // Traverse left and right subtrees with updated positions
    dfs(node->left, row + 1, col - 1, nodes);
    dfs(node->right, row + 1, col + 1, nodes);
}

// Main function for vertical order traversal
vector<vector<int>> verticalTraversal(TreeNode* root) {
    // Map to hold nodes grouped by columns
    map<int, priority_queue<tuple<int, int>, vector<tuple<int, int>>, greater<>>> nodes;

    // Perform DFS
    dfs(root, 0, 0, nodes);

    // Prepare the result
    vector<vector<int>> result;
    for (auto& [col, pq] : nodes) {
        vector<int> colNodes;
        while (!pq.empty()) {
            colNodes.push_back(get<1>(pq.top()));
            pq.pop();
        }
        result.push_back(colNodes);
    }
    return result;
}

};